src/metabase/lib/order_by.cljc

```diff
@@ -0,0 +1,74 @@
+(ns metabase.lib.order-by
+  (:require
+   [clojure.spec.alpha :as s]
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.field :as lib.field]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.order-by :as lib.schema.order-by]
+   [metabase.lib.util :as lib.util]
+   [metabase.util.malli :as mu]))
+
+(defmulti ^:private ->order-by-clause
+  {:arglists '([query stage-number x])}
+  (fn [_query _stage-number x]
+    (lib.dispatch/dispatch-value x)))
+
+(defmethod ->order-by-clause :asc
+  [_query _stage-number clause]
+  (lib.options/ensure-uuid clause))
+
+(defmethod ->order-by-clause :desc
+  [_query _stage-number clause]
+  (lib.options/ensure-uuid clause))
+
+;;; by default, try to convert `x` to a Field clause and then order by `:asc`
+(defmethod ->order-by-clause :default
+  [query stage-number x]
+  (let [field-clause (lib.field/field query stage-number x)]
+    (lib.options/ensure-uuid [:asc field-clause])))
+
+(defn order-by-clause
+  "Create an order-by clause independently of a query, e.g. for `replace` or whatever."
+  ([x]
+   (fn [query stage-number]
+     (order-by-clause query stage-number x)))
+  ([query stage-number x]
+   (->order-by-clause query stage-number x)))
+
+(mu/defn ^:private with-direction :- ::lib.schema.order-by/order-by
+  "Update the direction of an order by clause."
+  [clause    :- ::lib.schema.order-by/order-by
+   direction :- ::lib.schema.order-by/direction]
+  (assoc (vec clause) 0 direction))
+
+(s/def ::order-by-args
+  (s/cat
+   :query        map?
+   :stage-number (s/? integer?)
+   :x            any?
+   :direction    (s/? #{:asc :desc})))
```

Comment:
Yeah that's a good point. I think we can just disallow `:asc` or `:desc` as `x` since those don't make any sense at all anyway. The schema for `x` could definitely be more strict but I didn't know exactly what to put in there so I left it relatively open for now, we can always tweak it later.

Reply:
I ended up removing this spec anyway and just making `order-by` a normal multi-arity function